### PR TITLE
MSSQL states

### PR DESCRIPTION
**Pull request by pakdel**
### What does this PR do?
Added MSSQL states
Also added minor features to mssql execution module

### Tests written?
No

### CODE REVIEW

**Comment by pakdel** [comment]
Deprecates https://github.com/saltstack/salt/pull/42460

**Comment by pakdel** [comment]
@cachedout Could you please review this PullRequest?
This is a superset of https://github.com/saltstack/salt/pull/42316

**Comment by cachedout** [comment]
@pakdel I'm confused. You say this deprecates #42460 but that's this PR. :]

**Comment by pakdel** [comment]
sorry @cachedout 
That was a copy / paste error
I meant #42316 , like the comment after it

File: salt/modules/mssql.py [hunk 1]
```diff
@@ -25,6 +25,14 @@
 from __future__ import absolute_import
 from json import JSONEncoder, loads
 
+import sys
+if sys.version_info < (3,):
```
Comment (cachedout) [inline salt/modules/mssql.py]: We prefer to do this with `six.PY3` instead.

File: salt/modules/mssql.py [hunk 2]
```diff
@@ -250,23 +300,97 @@ def login_exists(login, **kwargs):
         return 'Could not find the login: {0}'.format(e)
 
 
-def user_exists(username, **kwargs):
+def login_create(login, new_login_password=None, new_login_domain=None, new_login_roles=None, new_login_options=None, **kwargs):
```
Comment (cachedout) [inline salt/modules/mssql.py]: Why change the name of this function?

Comment (pakdel) [inline salt/modules/mssql.py]: This is a diff issue.
`login_create` is a new function
`user_exists` is right after this function

Comment (cachedout) [inline salt/modules/mssql.py]: Ah, cool. Sounds good then.

File: salt/modules/mssql.py (hunk 1 above)
Comment (pakdel) [inline salt/modules/mssql.py]: Done

File: salt/modules/mssql.py [hunk 3]
```diff
@@ -195,19 +229,28 @@ def role_create(role, owner=None, **kwargs):
 
         salt minion mssql.role_create role=product01 owner=sysdba
     '''
+    if not grants:
```
Comment (cachedout) [inline salt/modules/mssql.py]: We need to modify the documentation here to show an example of using grants and indicate that if grants are to be passed in, they should be a list.

File: salt/modules/mssql.py [hunk 4]
```diff
@@ -232,7 +275,7 @@ def role_remove(role, **kwargs):
         return 'Could not create the role: {0}'.format(e)
 
 
-def login_exists(login, **kwargs):
+def login_exists(login, domain=None, **kwargs):
```
Comment (cachedout) [inline salt/modules/mssql.py]: We also need to modify the function documentation below to show that this should be passed a string, or perhaps even better just make `''` the default as well.

Comment (pakdel) [inline salt/modules/mssql.py]: Done

File: salt/modules/mssql.py (hunk 3 above)
Comment (cachedout) [inline salt/modules/mssql.py]: As soon as this requested change is made, I think we should be able to merge this in.

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by pakdel** [comment]
@cachedout I guess "Go Go Jenkins!" was supposed to merge this, correct?

**Review by cachedout — APPROVED** [review]

**Pull request merged**